router/handler_link/bind.go

```diff
@@ -0,0 +1,127 @@
+package handler_link
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/channel"
+	"github.com/openziti/channel/latency"
+	"github.com/openziti/channel/protobufs"
+	"github.com/openziti/fabric/pb/ctrl_pb"
+	"github.com/openziti/fabric/router/forwarder"
+	metrics2 "github.com/openziti/fabric/router/metrics"
+	"github.com/openziti/fabric/router/xgress"
+	"github.com/openziti/fabric/router/xlink"
+	"github.com/openziti/fabric/trace"
+	"github.com/openziti/foundation/metrics"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+func NewBindHandlerFactory(c xgress.CtrlChannel, f *forwarder.Forwarder, fo *forwarder.Options, mr metrics.Registry) *bindHandlerFactory {
+	return &bindHandlerFactory{
+		ctrl:             c,
+		forwarder:        f,
+		forwarderOptions: fo,
+		metricsRegistry:  mr,
+	}
+}
+
+type bindHandlerFactory struct {
+	ctrl             xgress.CtrlChannel
+	forwarder        *forwarder.Forwarder
+	forwarderOptions *forwarder.Options
+	metricsRegistry  metrics.Registry
+}
+
+func (self *bindHandlerFactory) NewBindHandler(link xlink.Xlink, latency bool, listenerSide bool) channel.BindHandler {
+	return &bindHandler{
+		bindHandlerFactory: self,
+		xlink:              link,
+		trackLatency:       latency,
+		listenerSide:       listenerSide,
+	}
+}
+
+type bindHandler struct {
+	*bindHandlerFactory
+	xlink        xlink.Xlink
+	trackLatency bool
+	listenerSide bool
+}
+
+func (self *bindHandler) BindChannel(binding channel.Binding) error {
+	ch := binding.GetChannel()
+	if self.listenerSide {
+		if err := self.verifyLink(self.xlink, ch); err != nil {
+			return err
+		}
+	}
+
+	closeNotify := make(chan struct{})
+
+	binding.GetChannel().SetLogicalName("l/" + self.xlink.Id().Token)
+	binding.SetUserData(self.xlink.Id().Token)
+	binding.AddCloseHandler(newCloseHandler(self.xlink, self.ctrl, self.forwarder, closeNotify))
+	binding.AddErrorHandler(newErrorHandler(self.xlink, self.ctrl))
+	binding.AddTypedReceiveHandler(newPayloadHandler(self.xlink, self.forwarder))
+	binding.AddTypedReceiveHandler(newQueuingAckHandler(self.xlink, self.forwarder, closeNotify))
+	binding.AddTypedReceiveHandler(&latency.LatencyHandler{})
+	binding.AddTypedReceiveHandler(newControlHandler(self.xlink, self.forwarder, closeNotify))
+	binding.AddPeekHandler(metrics2.NewChannelPeekHandler(self.xlink.Id().Token, self.forwarder.MetricsRegistry()))
+	binding.AddPeekHandler(trace.NewChannelPeekHandler(self.xlink.Id().Token, ch, self.forwarder.TraceController(), trace.NewChannelSink(self.ctrl.Channel())))
+
+	if self.trackLatency {
+		latencyMetric := self.metricsRegistry.Histogram("link." + self.xlink.Id().Token + ".latency")
```

Comment:
is there any way to embed `self.xlink.Id().Token` as a tag or anything else? aren't unbounded names like this 'frowned on'?

Reply:
Metric names have to be unique. This gets transformed when it goes through the events subsystem into a metric named `link.latency` with tags for the source and target routers.